Glue metastore statistics integration

GaruGaru opened the pull request:
Integrate aws glue statistics api in glue metastore, solves #5072.

There are some points that may be worth reviewing: 

* Since the aws sdk has limits for read / write operations in order to improve performance with an high number of columns all write / read operations are executed by a dedicated thread pool, in order to pass multiple Executors i've introduced the `ForGlueColumnStatisticsWrite` and `ForGlueColumnStatisticsRead` annotations but I'm not sure if that is a guice's best practice.

*  The `getSupportedColumnStatistics` is calling `ThriftMetastoreUtil#getSupportedColumnStatistics` we may want to duplicate the code to avoid unwanted side effects.

*  There is some specific code implemented in `DefaultGlueColumnStatisticsProvider` to match the Hive implementation that clear statistics on update request when the requested columns are empty, I'm not sure if we have have to keep the behaviour in the Glue implementation or is just needed for the test. 

* I modified the GlueColumnStatisticsProvider in order to pass `Table` and `Partition` instead of the Glue object `TableInput`, `PartitionInput` because these object weren't able to provide all the needed data to create request for the glue api

svarosky90 commented:
Hi Trinoers,
any idea about when you expect to merge this pull request in a new version of trino?
Thank you.

hashhar commented:
@GaruGaru Can you please rebase your changes on top of `master` to accomodate the rename?

You might find https://github.com/trinodb/trino/wiki/Developer-Migration useful.

hashhar commented on presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/glue/converter/GlueStatConverter.java:
```diff
@@ -0,0 +1,315 @@
+            ColumnStatistics catalogColumnStatistics = new ColumnStatistics();
+
+            catalogColumnStatistics.setColumnName(columnName);
+            catalogColumnStatistics.setColumnType(columnType.toString());
```
Are the various allowed column types documented somewhere? Can there be a case where a HiveType isn't supported by Glue?

hashhar replied:
Same in `toGlueColumnStatistics`.

GaruGaru replied:
The switch case follow the aws glue api reference, if the type is not supported a `TrinoException` with `HIVE_INVALID_METADATA` is thrown in both `toGlueColumnStatisticsData` and `fromGlueColumnStatistics`,
Do you think we can handle this better ? Should I also add the docs link as reference to the to/from methods ? 

api reference: https://docs.aws.amazon.com/glue/latest/dg/aws-glue-api-common.html#aws-glue-api-common-ColumnStatistics

hashhar replied:
This makes sense. Consider this resolved.

hashhar commented on presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/glue/converter/GlueStatConverter.java:
```diff
@@ -0,0 +1,315 @@
+    {
+        List<ColumnStatistics> catalogColumnStatisticsList = new ArrayList<>(prestoColumnStats.size());
+
+        prestoColumnStats.forEach((columnName, statistics) -> {
```
We can instead loop on `partition.getColumns` to avoid the lookup in `columnByName`.

I don't think there can be cases where `partition.getColumns` doesn't match columns in `prestoColumnStats`. Please correct me if I'm wrong.

hashhar commented on presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/glue/converter/GlueStatConverter.java:
```diff
@@ -0,0 +1,315 @@
+        throw new PrestoException(HIVE_INVALID_METADATA, "Invalid column statistics data: " + catalogColumnStatisticsData);
+    }
+
+    private static ColumnStatisticsData getGlueColumnStatisticsData(HiveColumnStatistics statistics, HiveType columnType, OptionalLong rowCount)
```
```suggestion
    private static ColumnStatisticsData toGlueColumnStatisticsData(HiveColumnStatistics statistics, HiveType columnType, OptionalLong rowCount)
```

nit: Rename the method to follow the `fromXXX` and `toXXX` pattern.

hashhar commented on presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/glue/converter/GlueStatConverter.java:
```diff
@@ -0,0 +1,315 @@
+    private static Optional<LocalDate> dateToLocalDate(Date date)
+    {
+        long daysSinceEpoch = date.getTime() / DAY_TO_MILLISECOND_FACTOR;
+        return Optional.of(LocalDate.ofEpochDay(daysSinceEpoch));
```
Someone more familiar with datetime gotchas should take a look at this.

GaruGaru replied:
We may use  ```LocalDate.ofInstant``` but I'm not sure how to handle the ZoneId parameter

hashhar commented on presto-hive/src/test/java/io/prestosql/plugin/hive/metastore/glue/TestHiveGlueMetastore.java:
```diff
@@ -152,34 +154,59 @@ public void testRenameTable()
     {
-        // column statistics are not supported by Glue
+        // this test is not really meaningful for Glue, for numeric columns, min/max stat should always be available
+        // The only exception is when all rows are NULLs. Glue requires min/max to be set.
```
Can we then change this to an `assertThat(super.testUpdateTableColumnStatisticsEmptyOptionalFields()).throws(xxx)`?

That way if Glue's behaviour changes in future we'll know that we can/need to change our impl.

Same for other commented out tests.

GaruGaru replied:
Agree, but we should also be careful with this since glue is returning http status code `500` so this case is not handled by their api yet, I've added some specific check on the error message to address this

hashhar commented on pom.xml:
```diff
@@ -48,7 +48,7 @@
         <dep.airlift.version>201</dep.airlift.version>
         <dep.packaging.version>${dep.airlift.version}</dep.packaging.version>
-        <dep.aws-sdk.version>1.11.749</dep.aws-sdk.version>
+        <dep.aws-sdk.version>1.11.901</dep.aws-sdk.version>
```
Nevermind, seems like we are already updated to an even newer version on `master`.

GaruGaru replied:
yes after rebase we are running on `1.11.921`

hashhar commented on presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/glue/DefaultGlueColumnStatisticsProvider.java:
```diff
@@ -0,0 +1,227 @@
+    @Override
+    public Set<ColumnStatisticType> getSupportedColumnStatistics(Type type)
+    {
+        return ThriftMetastoreUtil.getSupportedColumnStatistics(type);
```
Nevermind, I saw that you are picking the impl. based on the config toggle so that impls. don't need to be aware if things are enabled or disabled. Nice.

GaruGaru replied:
Yes, I'm not sure if having a No op implementation is a common pattern in the codebase, we may also change the default implementation to use a toggle flag if you think it's more appropriate

hashhar replied:
We do have some no-op implementations in other places in the code. It looks fine to me.

But I'll let the maintainers take the call on this one.

cc: @findepi @electrum

findepi replied:
We did have a no-op implementation because stats for glue was extensions point. It was needed until Glue provided first-class stats support.
We should maintain the extension point for a while, before we can remove it.
So the DefaultGCSP can change its' behavior, but should not be used uncondintionally. cc @sopel39 

Also, we need a safety toggle to be able to ignore stats, should users run into any problems.
@GaruGaru can you please confirm `hive.table-statistics-enabled` can be used to skip getting stats from Glue?

GaruGaru replied:
yes disabling `hive.table-statistics-enabled` will use the current No op implementation

GaruGaru replied:
In order to enable/disable the statistics get/update I've introduced the `hive.metastore.glue.statistics-enabled` flag, do you think we can use `hive.table-statistics-enabled` directly or we should check them both ?

findepi replied:
One flag should be enough. You cannot use 2 metastores at the same time anyway.

GaruGaru replied:
I've removed `hive.metastore.glue.statistics-enabled` in favour of `hive.table-statistics-enabled`

hashhar commented on presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/glue/DefaultGlueColumnStatisticsProvider.java:
```diff
@@ -0,0 +1,227 @@
+    {
+        final HiveBasicStatistics tableStats = getHiveBasicStatistics(table.getParameters());
+
+        if (columnStatistics.isEmpty()) {
```
I now see that this class is used only when stats are enabled so the condition to switch on makes sense.

Still, this method behaves differently compared to the one for ThriftHiveMetastore. The impl I'm suggesting would preserve semantics.

GaruGaru replied:
Agree, the code with the boolean flag / check was written that way to preserve the ThriftHiveMetastore behaviour but keeping the same semantic is the right thing to do here in both `updatePartitionStatistics`and  `updateTableColumnStatistics`.

This change is likely to impact the performance of the implementation (glue api call) but from my understanding the update statistics process is not that latency sensitive

hashhar commented on presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/glue/GlueColumnStatisticsProvider.java:
```diff
@@ -32,7 +30,7 @@
+    void updateTableColumnStatistics(Table table, Map<String, HiveColumnStatistics> columnStatistics);
 
-    void updatePartitionStatistics(PartitionInput partition, Map<String, HiveColumnStatistics> columnStatistics);
+    void updatePartitionStatistics(Partition partition, Map<String, HiveColumnStatistics> columnStatistics, boolean forPartitionCreation);
```
Does the change suggested in `updateTableColumnStatistics` allow us to remove this `boolean` control variable?

I'm not in favour of using boolean arguments to decide method behaviour.

GaruGaru replied:
Totally agree with removing the boolean flag, the proposed change should address this

hashhar reviewed (commented):
First pass. Thanks for contributing this.

Some high level comments:

1. Every use of `getFutureValue` should make sure to handle exceptions and return the proper error code by wrapping in a PrestoException.
2. I've not reviewed `GlueStatConverter` in depth yet.

Please rebase your branch on top of master to accomodate the rename to Trino.
I have a rebased version of your branch at https://github.com/hashhar/trino/tree/metastore-glue-statistics.

GaruGaru commented on presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/glue/DefaultGlueColumnStatisticsProvider.java:
```diff
@@ -0,0 +1,227 @@
+public class DefaultGlueColumnStatisticsProvider
+        implements GlueColumnStatisticsProvider
+{
+    private static final int GLUE_COLUMN_READ_STAT_PAGE_SIZE = 100;
```
Right, I've added the explanation with the aws api reference link

GaruGaru commented on presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/glue/DefaultGlueColumnStatisticsProvider.java:
```diff
@@ -0,0 +1,227 @@
+
+        final HiveBasicStatistics tableStatistics = getHiveBasicStatistics(table.getParameters());
+
+        final ImmutableMap.Builder<String, HiveColumnStatistics> columnStatsMapBuilder = ImmutableMap.builder();
```
Nice catch, I've changed the implementation as you suggested

GaruGaru commented on presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/glue/DisabledGlueColumnStatisticsProvider.java:
```diff
@@ -51,15 +49,15 @@
 
     @Override
-    public void updateTableColumnStatistics(TableInput table, Map<String, HiveColumnStatistics> columnStatistics)
+    public void updateTableColumnStatistics(Table table, Map<String, HiveColumnStatistics> columnStatistics)
```
I agree, We may have a more defined separation between the packages but we need models defined by `Table`, `Column ,`Partition` classes. 
We can either duplicates the attributes to some Glue specific wrappers or create some shared definitions but that may be risky in case of hive/glue api changes

hashhar replied:
This is fine to keep as is. If there is some breakage in future (unlikely) then we can take a look at extracting a wrapper.

findepi commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/glue/TestHiveGlueMetastore.java:
```diff
@@ -152,34 +154,59 @@ public void testRenameTable()
     {
-        // column statistics are not supported by Glue
+        // this test is not really meaningful for Glue, for numeric columns, min/max stat should always be available
+        // The only exception is when all rows are NULLs. Glue requires min/max to be set.
```
`throw new SkipException` would be more appropriate, but i'd still want to understand better what is supposed to be tested here and cannot be

GaruGaru replied:
This is kinda strange, from my understanding here the test try to perform a statistics update for all the columns but without min/max values.
Min / Max values are actually optional according to the [docs](https://docs.aws.amazon.com/glue/latest/webapi/API_LongColumnStatisticsData.html) .
But the Glue api answer with an `Status Code: 500; Error Code: InternalServiceException` Without giving much context. 
Requests were all the statistics (min, max, nulls ...) are specified are completing correctly

losipiuk replied:
@GaruGaru knowing the deficiency of Glue can we make the code write minimal/maximal value for a type if actual min/max is not present in `HiveColumnStatistics`?
So for BIGINT we would write `-2^63`/`2^63 - 1` for INTEGER `-2^31``/`2^31 - 1` ... and so on.

losipiuk replied:
@GaruGaru has this one been addressed?

GaruGaru commented on presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/glue/DefaultGlueColumnStatisticsProvider.java:
```diff
@@ -0,0 +1,227 @@
+
+        final ImmutableList.Builder<ColumnStatistics> columnStatsBuilder = ImmutableList.builder();
+        for (CompletableFuture<GetColumnStatisticsForTableResult> future : getStatsFutures.build()) {
+            final GetColumnStatisticsForTableResult tableColumnsStats = MoreFutures.getFutureValue(future, PrestoException.class);
```
I solved this but at the moment I'm catching `AWSGlueException`, should the catch clause also check for more generic exceptions ?

hashhar commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/DefaultGlueColumnStatisticsProvider.java:
```diff
@@ -92,17 +98,14 @@ public DefaultGlueColumnStatisticsProvider(AWSGlueAsync glueClient, String catal
-            columnStatsBuilder.addAll(tableColumnsStats.getColumnStatisticsList());
+            final GetColumnStatisticsForTableResult tableColumnsStats = MoreFutures.getFutureValue(future, TrinoException.class);
+            tableColumnsStats.getColumnStatisticsList()
+                    .forEach(col -> columnStatsMapBuilder.put(col.getColumnName(), GlueStatConverter.fromGlueColumnStatistics(col.getStatisticsData(), tableStatistics.getRowCount())));
```
Add static import for `fromGlueColumnStatistics`.

Also in `getPartitionColumnStatistics`.

hashhar commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/converter/GlueStatConverter.java:
```diff
@@ -279,7 +279,7 @@ private static ColumnStatisticsData getGlueColumnStatisticsData(HiveColumnStatis
                 break;
             default:
-                throw new PrestoException(HIVE_INVALID_METADATA, "Invalid column statistics type: " + statistics);
+                throw new TrinoException(HIVE_INVALID_METADATA, "Invalid column statistics type: " + statistics);
```
How does the `statistics` object look when logged?

Would it be more useful to just log `((PrimitiveTypeInfo) typeInfo).getPrimitiveCategory()`?

hashhar commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/glue/TestHiveGlueMetastore.java:
```diff
@@ -179,16 +179,18 @@ public void testUpdateTableColumnStatistics()
             throws Exception
     {
-        // this test is not really meaningful for Glue, for numeric columns, min/max stat should always be available
-        // The only exception is when all rows are NULLs. Glue requires min/max to be set.
```
I'd keep the comment about why this fails. `Glue requires min/max to be set` or more accurately `Glue doesn't have any optional fields in statistics`.

But this doesn't match the documentation from AWS somehow. They mention min/max as `Required: No`.
If it's possible can you take a look at the response returned from AWS by enabling their debug logging as decribed at https://docs.aws.amazon.com/sdk-for-java/v1/developer-guide/java-dg-logging.html#sdk-net-logging-request-response.

hashhar commented:
The only high level things needing attention now are:

1. The difference in `updateTableColumnStatistics` semantics between the ThriftHiveMetastore and GlueHiveMetastore.
  The Thrift impl acts more like an `overwrite`. i.e. if existing stats are for columns `a, b, c` and the updated stats are for columns `a, d` then the final stats become `a, d`.
  The Glue impl acts more like an `append`. Given same inputs above the final stats become `a, b, c, d`.
  This has the side effect of using a boolean to decide if we want to drop stats for some columns or overwrite them.

2. The exception handling from the executors used to make the read and write calls to Glue API. My expectation is that any failing future should cause the entire method to return a `TrinoException(HIVE_METASTORE_ERROR, e)`. I'm not sure what's the cleanest way to do it. The current wrapping inside a large try ... catch does seem suboptimal.

3. Proper cancellation of futures when one of them fails. This is missing today and can lead to some subtle bugs where table stats can be inconsistent since the update call failed in one future but passed in other 3.

4. The conversions of date/time in `GlueStatConverter`. It looks okay to me but someone should take a look if the conversions aren't susceptible to issues like gaps in time. Or we can add unit tests for the function to make sure stats conversion is done correctly?

cc: @trinodb/maintainers for some inputs.

GaruGaru commented:
@hashhar @findepi  thank you for the feedbacks ! I'm working to address all the issues. 
I'm trying to clarify the `Glue internal server error` on empty statistics to understand how we should handle `testUpdateTableColumnStatisticsEmptyOptionalFields` and `testUpdatePartitionColumnStatisticsEmptyOptionalFields`

Regarding your points:

1. I've changed the implementation to match the semantic of `ThriftHiveMetastore`, I've managed to remove the boolean flag but since the update method is also called by `addPartition` before the actual partition creation I had to add another api call to Glue in order to check if the partition is present before applying the new logic. 
2. Maybe splitting the implementation in different methods may help for a better error handling ? 
3. Do you think we should implement some kind of rollback strategy as implemented in `SemiTransactionalHiveMetastore` ? What do you think is the optimal behaviour in case of partial failure ? Should we just write back the previously available statistics for all the columns ?

losipiuk commented:
@GaruGaru Please rebase and squash in the fixup commits to make the review easier. I will do the first pass looking at overall PR as going commit by commit is hard to do right now.

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/GlueHiveMetastore.java:
```diff
@@ -185,10 +189,16 @@ public GlueHiveMetastore(
         this.assumeCanonicalPartitionKeys = glueConfig.isAssumeCanonicalPartitionKeys();
         this.tableFilter = requireNonNull(tableFilter, "tableFilter is null");
+        this.enableColumnStatistics = hiveConfig.isTableStatisticsEnabled();
+        if (this.enableColumnStatistics) {
```
please setup that in the module:
```java
        installModuleIf(
                HiveConfig.class,
                HiveConfig::isTableStatisticsEnabled,
                innerBinder -> innerBinder.bind(GlueColumnStatisticsProvider.class).to(DefaultGlueColumnStatisticsProvider.class).in(Scopes.SINGLETON),
                innerBinder -> innerBinder.bind(GlueColumnStatisticsProvider.class).to(DisabledGlueColumnStatisticsProvider.class).in(Scopes.SINGLETON));

```
should do

losipiuk replied:
Oh I see that it would require some more refactor regarding stats so those are guice bound too. I guess it can be done as a followup.

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/GlueHiveMetastoreConfig.java:
```diff
@@ -40,6 +40,8 @@
     private Optional<String> catalogId = Optional.empty();
     private int partitionSegments = 5;
     private int getPartitionThreads = 20;
+    private int readStatisticsThreads = 1;
```
Do we need those configuration parameters? Do we expect that we would need to bump it? If not I would rather not have those. The less config params the better.

GaruGaru replied:
I agree with having as few params as possible but this parameters are useful for performance tuning when working with tables with an high amount of columns, also making the concurrency configurable should help users to deal with the AWS Glue rate limits.

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/GlueMetastoreModule.java:
```diff
@@ -46,10 +47,7 @@
-        newOptionalBinder(binder, GlueColumnStatisticsProvider.class)
-                .setDefault().to(DisabledGlueColumnStatisticsProvider.class).in(Scopes.SINGLETON);
-
+        configBinder(binder).bindConfig(HiveConfig.class);
```
It feels to me that you should not need this one as it is bind in `HIveModule` which should be always present.

GaruGaru replied:
Do we have a test that verify this ?

losipiuk replied:
I do not believe we have. AFAIK (and I could not find any) we have no proper product-tests coverage for Trino run agains AWS.
cc: @findepi

findepi replied:
Correct, we have only non-product tests, like `TestHiveGlueMetastore`
this is tracked by https://github.com/trinodb/trino/issues/5426

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/GlueMetastoreModule.java:
```diff
@@ -59,6 +57,17 @@ protected void setup(Binder binder)
                 .to(GlueHiveMetastore.class)
                 .in(Scopes.SINGLETON);
+
+        binder.bind(HiveMetastore.class)
+                .annotatedWith(ForGlueColumnStatisticsRead.class)
+                .to(GlueHiveMetastore.class)
+                .in(Scopes.SINGLETON);
+
+        binder.bind(HiveMetastore.class)
+                .annotatedWith(ForGlueColumnStatisticsWrite.class)
+                .to(GlueHiveMetastore.class)
+                .in(Scopes.SINGLETON);
```
those are not used anywhere.

GaruGaru replied:
Used in GlueHiveMetastore's constructor

losipiuk replied:
The `Executors` (which are bound below) are used in `GlueHiveMetastore`. This one is adding binging for `HiveMetastore` (annotated with `ForGlueColumnStatisticsRead` and `ForGlueColumnStatisticsWrite`).

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/converter/GlueStatConverter.java:
```diff
@@ -0,0 +1,315 @@
+        List<ColumnStatistics> catalogColumnStatisticsList = new ArrayList<>(trinoColumnStats.size());
+
+        trinoColumnStats.forEach((columnName, statistics) -> {
+            final Optional<Column> column = columnByName(partition, columnName);
```
drop `final` modifiers on non-fields. We do not have a convention in Trino to use them.

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/GlueMetastoreModule.java:
```diff
@@ -75,7 +84,33 @@ public Executor createExecutor(CatalogName catalogName, GlueHiveMetastoreConfig
+    @ForGlueColumnStatisticsRead
+    public Executor createStatisticsReadExecutor(CatalogName catalogName, GlueHiveMetastoreConfig hiveConfig)
+    {
+        if (hiveConfig.getReadStatisticsThreads() == 1) {
+            return directExecutor();
+        }
+        return new BoundedExecutor(
+                newCachedThreadPool(daemonThreadsNamed("hive-glue-statistics-read-%s")),
+                hiveConfig.getReadStatisticsThreads());
```
extract helper method to create executor given number of threads and name. All three look exactly the same.

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/DefaultGlueColumnStatisticsProvider.java:
```diff
@@ -0,0 +1,270 @@
+            final HiveBasicStatistics tableStatistics = getHiveBasicStatistics(table.getParameters());
+            final ImmutableMap.Builder<String, HiveColumnStatistics> columnStatsMapBuilder = ImmutableMap.builder();
+            for (CompletableFuture<GetColumnStatisticsForTableResult> future : getStatsFutures.build()) {
+                final GetColumnStatisticsForTableResult tableColumnsStats = MoreFutures.getFutureValue(future, TrinoException.class);
```
static import `MoreFutures.getFutureValue`

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/DefaultGlueColumnStatisticsProvider.java:
```diff
@@ -0,0 +1,270 @@
+                return glueClient.getColumnStatisticsForTable(request);
+            }, readExecutor)));
+
+            final HiveBasicStatistics tableStatistics = getHiveBasicStatistics(table.getParameters());
+            final ImmutableMap.Builder<String, HiveColumnStatistics> columnStatsMapBuilder = ImmutableMap.builder();
+            for (CompletableFuture<GetColumnStatisticsForTableResult> future : getStatsFutures.build()) {
+                final GetColumnStatisticsForTableResult tableColumnsStats = MoreFutures.getFutureValue(future, TrinoException.class);
+                tableColumnsStats.getColumnStatisticsList()
+                        .forEach(col -> columnStatsMapBuilder.put(col.getColumnName(), GlueStatConverter.fromGlueColumnStatistics(col.getStatisticsData(), tableStatistics.getRowCount())));
+            }
+
+            return columnStatsMapBuilder.build();
```
```suggestion
            HiveBasicStatistics tableBasicStatistics = getHiveBasicStatistics(table.getParameters());
            ImmutableMap.Builder<String, HiveColumnStatistics> tableColumnStatistics = ImmutableMap.builder();
            for (CompletableFuture<GetColumnStatisticsForTableResult> future : getStatsFutures.build()) {
                GetColumnStatisticsForTableResult tableColumnsStats = getFutureValue(future, TrinoException.class);
                for (ColumnStatistics singleColumnStatistics : tableColumnsStats.getColumnStatisticsList()) {
                    tableColumnStatistics.put(
                            singleColumnStatistics.getColumnName(),
                            fromGlueColumnStatistics(singleColumnStatistics.getStatisticsData(), tableBasicStatistics.getRowCount()));
                }
            }

            return tableColumnStatistics.build();
```

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/DefaultGlueColumnStatisticsProvider.java:
```diff
@@ -0,0 +1,270 @@
+                return glueClient.getColumnStatisticsForPartition(request);
+            }, readExecutor)));
+
+            final HiveBasicStatistics tableStatistics = getHiveBasicStatistics(partition.getParameters());
```
Please restructure in similar way as above. I think it make the code more readable.

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/DefaultGlueColumnStatisticsProvider.java:
```diff
@@ -0,0 +1,270 @@
+public class DefaultGlueColumnStatisticsProvider
+        implements GlueColumnStatisticsProvider
+{
+    // Write limit for AWS Glue API GetColumnStatisticsForPartition
```
```suggestion
    // Read limit for AWS Glue API GetColumnStatisticsForPartition
```

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/DefaultGlueColumnStatisticsProvider.java:
```diff
@@ -0,0 +1,270 @@
+
+            ImmutableList.Builder<CompletableFuture<GetColumnStatisticsForTableResult>> getStatsFutures = ImmutableList.builder();
+
+            columnChunks.forEach(partialColumns -> getStatsFutures.add(CompletableFuture.supplyAsync(() -> {
```
can you please check if `stream.map.collect` would not look nicer?

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/DefaultGlueColumnStatisticsProvider.java:
```diff
@@ -0,0 +1,270 @@
+                                        .withColumnName(column.getName())), writeExecutor))
+                        .collect(Collectors.toUnmodifiableList());
+
+                for (CompletableFuture<Void> writeFuture : writeFutures) {
+                    MoreFutures.getFutureValue(writeFuture);
+                }
```
maybe
```
                getFutureValue(allOf(writeFutures.toArray(CompletableFuture[]::new)));
```

I am not super convinced. I would be more if there was `allOf` which takes a list ....

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/DefaultGlueColumnStatisticsProvider.java:
```diff
@@ -0,0 +1,270 @@
+
+            if (columnStatistics.isEmpty()) {
+                final List<CompletableFuture<Void>> writeFutures = Stream.concat(table.getDataColumns().stream(), table.getPartitionColumns().stream())
+                        .map(column -> CompletableFuture.runAsync(() -> glueClient.deleteColumnStatisticsForTable(
```
static import `runAsync`

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/DefaultGlueColumnStatisticsProvider.java:
```diff
@@ -0,0 +1,270 @@
+                                        .withDatabaseName(table.getDatabaseName())
+                                        .withTableName(table.getTableName())
+                                        .withColumnName(column.getName())), writeExecutor))
+                        .collect(Collectors.toUnmodifiableList());
```
static import

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/DefaultGlueColumnStatisticsProvider.java:
```diff
@@ -0,0 +1,270 @@
+                }
+            }
+
+            final List<ColumnStatistics> columnStats = GlueStatConverter.toGlueColumnStatistics(table, columnStatistics, tableStats.getRowCount());
```
static import

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/DefaultGlueColumnStatisticsProvider.java:
```diff
@@ -0,0 +1,270 @@
+                for (CompletableFuture<Void> writeFuture : writeFutures) {
+                    MoreFutures.getFutureValue(writeFuture);
+                }
+            }
```
missing return?

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/DefaultGlueColumnStatisticsProvider.java:
```diff
@@ -0,0 +1,270 @@
+                                    .withDatabaseName(table.getDatabaseName())
+                                    .withTableName(table.getTableName())
+                                    .withColumnStatisticsList(columnChunk))))
+                    .collect(Collectors.toUnmodifiableList());
```
static import

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/DefaultGlueColumnStatisticsProvider.java:
```diff
@@ -0,0 +1,270 @@
+            final List<ColumnStatistics> columnStats = GlueStatConverter.toGlueColumnStatistics(table, columnStatistics, tableStats.getRowCount());
+            final List<List<ColumnStatistics>> columnChunks = Lists.partition(columnStats, GLUE_COLUMN_WRITE_STAT_PAGE_SIZE);
+
+            final List<CompletableFuture<Void>> writeChunkFutures = columnChunks.stream().map(columnChunk -> CompletableFuture.runAsync(
```
static import `runAsync`

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/DefaultGlueColumnStatisticsProvider.java:
```diff
@@ -0,0 +1,270 @@
+            final List<List<ColumnStatistics>> columnChunks = Lists.partition(columnStats, GLUE_COLUMN_WRITE_STAT_PAGE_SIZE);
+
+            final List<CompletableFuture<Void>> writePartitionStatsFutures = columnChunks.stream()
+                    .map(column ->
```
name the variables `columnChunk`

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/DefaultGlueColumnStatisticsProvider.java:
```diff
@@ -0,0 +1,270 @@
+            final Map<String, HiveColumnStatistics> currentColumnStatistics = partitionExists ? this.getPartitionColumnStatistics(partition) : Collections.emptyMap();
+            final Set<String> removedStatistics = difference(currentColumnStatistics.keySet(), updatedColumnStatistics.keySet());
+
+            final List<CompletableFuture<Void>> writeFutures = removedStatistics.stream()
```
Why don't we need analogous code in `updateTableColumnStatistics`?

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/DefaultGlueColumnStatisticsProvider.java:
```diff
@@ -0,0 +1,270 @@
+
+    private List<String> getAllColumns(Table table)
+    {
+        final List<String> allColumns = new ArrayList<>(table.getDataColumns().size() + table.getPartitionColumns().size());
```
Use ImmutableListBuilder like everywhere

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/converter/GlueStatConverter.java:
```diff
@@ -0,0 +1,315 @@
+        List<ColumnStatistics> catalogColumnStatisticsList = new ArrayList<>(trinoColumnStats.size());
+
+        trinoColumnStats.forEach((columnName, statistics) -> {
+            final Optional<Column> column = columnByName(partition, columnName);
```
this is O(N^2) which with 1000 columns will start to be substantial cost. Pre-build String->Column map

losipiuk replied:
Naa - it is only partition columns. So I guess we will not have 1000 of those :)

hashhar replied:
Still should be easy to remedy and there's no reason to avoid this unless it's more difficult than expected - https://github.com/trinodb/trino/pull/6178#discussion_r559353652

GaruGaru replied:
fixed

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/converter/GlueStatConverter.java:
```diff
@@ -0,0 +1,315 @@
+            HiveType columnType = table.getColumn(columnName).get().getType();
+            ColumnStatistics catalogColumnStatistics = new ColumnStatistics();
+
+            catalogColumnStatistics.setColumnName(columnName);
+            catalogColumnStatistics.setColumnType(columnType.toString());
+            ColumnStatisticsData catalogColumnStatisticsData = toGlueColumnStatisticsData(statistics, columnType, rowCount);
+            catalogColumnStatistics.setStatisticsData(catalogColumnStatisticsData);
+            catalogColumnStatistics.setAnalyzedTime(new Date());
```
can you extract common part for column and partition stats to helper method?

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/converter/GlueStatConverter.java:
```diff
@@ -0,0 +1,315 @@
+        throw new TrinoException(HIVE_INVALID_METADATA, "Invalid column statistics data: " + catalogColumnStatisticsData);
+    }
+
+    private static ColumnStatisticsData toGlueColumnStatisticsData(HiveColumnStatistics statistics, HiveType columnType, OptionalLong rowCount)
```
put this one below `columnByName`

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/converter/GlueToTrinoConverter.java:
```diff
@@ -93,7 +93,7 @@ public static Table convertTable(com.amazonaws.services.glue.model.Table glueTab
     }
 
-    private static Column convertColumn(com.amazonaws.services.glue.model.Column glueColumn)
+    public static Column convertColumn(com.amazonaws.services.glue.model.Column glueColumn)
```
seems not needed. revert

losipiuk commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/glue/TestHiveGlueMetastore.java:
```diff
@@ -152,34 +158,61 @@ public void testRenameTable()
+            throws Exception
     {
-        // column statistics are not supported by Glue
+        assertThatThrownBy(super::testUpdateTableColumnStatisticsEmptyOptionalFields)
```
this does not seem right? Is that a flow we can get into in real life?

losipiuk commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/glue/TestHiveGlueMetastore.java:
```diff
@@ -152,34 +158,61 @@ public void testRenameTable()
+            throws Exception
     {
-        // column statistics are not supported by Glue
+        assertThatThrownBy(super::testUpdatePartitionColumnStatisticsEmptyOptionalFields)
```
same here

losipiuk reviewed (commented):
Thanks! I put some suggestions and questions.
I did not review the converters part deeply.

hashhar commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/DefaultGlueColumnStatisticsProvider.java:
```diff
@@ -0,0 +1,272 @@
+            }
+
+            boolean partitionExists = partitionExists(partition);
+            Map<String, HiveColumnStatistics> currentColumnStatistics = partitionExists ? this.getPartitionColumnStatistics(partition) : Collections.emptyMap();
```
What happens if `getPartitionStatistics` is called for a non-existent partition?

There will be 1 API call on the not exists path but 2 on the exists path. Can we avoid this by using the exception thrown when a partition doesn't exist as a control variable instead of the boolean?

Lots of people run into Glue rate-limits so we should strive to reduce API calls to Glue.

GaruGaru replied:
When called on a non existing partition glue client throw an exception, we may use the exception to check if the partition is present as it will reduce the number of api calls.
In `getPartitionColumnStatistics` we should check for `EntityNotFoundException` raised by glue client and in that case return an empty partition statistics object, do you think if may be a good compromise between readability and api calls count ?

hashhar reviewed (commented):
Looks good to me now % existing comments.

I still need to check why the test stats values had to be changed.

hashhar commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/DefaultGlueColumnStatisticsProvider.java:
```diff
@@ -0,0 +1,266 @@
+                                            .withDatabaseName(table.getDatabaseName())
+                                            .withTableName(table.getTableName())
+                                            .withColumnName(column)))
+                    ).collect(toUnmodifiableList());
```
nit: move closing brackets to previous line. I wonder why checkstyle didn't complain though.

hashhar commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/DefaultGlueColumnStatisticsProvider.java:
```diff
@@ -0,0 +1,266 @@
+                    ).collect(toUnmodifiableList());
+
+            for (CompletableFuture<Void> writeFuture : writeFutures) {
+                getFutureValue(writeFuture);
```
Let's change all occurrences of loops over future where returned values are not checked into something like `getFutureValue(allOf(...))`? Or is that not equivalent?

losipiuk replied:
Yeah - please do that. So we acutally wait for all the futures to complete. And we are not leaving work still being processed in executors.

hashhar commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/DefaultGlueColumnStatisticsProvider.java:
```diff
@@ -0,0 +1,266 @@
+                                    .withTableName(partition.getTableName())
+                                    .withPartitionValues(partition.getValues())
+                                    .withColumnStatisticsList(columnChunk)), writeExecutor)
+                    ).collect(toUnmodifiableList());
```
nit: move closing brackets to previous line. I wonder why checkstyle didn't complain though.

hashhar commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/DefaultGlueColumnStatisticsProvider.java:
```diff
@@ -0,0 +1,266 @@
+                getFutureValue(writePartitionStatsFuture);
+            }
+
+            boolean partitionExists = partitionExists(partition);
```
Let's test for an exception to be thrown instead to avoid an API call. WDYT @losipiuk.

losipiuk replied:
If there is clear error code in exception with "partition not found" meaning I think we should use exception to avoid extra API call on success path.

hashhar commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/glue/TestHiveGlueMetastore.java:
```diff
@@ -152,34 +158,22 @@ public void testRenameTable()
-    public void testUpdatePartitionColumnStatistics()
-    {
-        // column statistics are not supported by Glue
+        assertThatThrownBy(super::testUpdateTableColumnStatisticsEmptyOptionalFields)
```
Let's add some comment about why this is failing? AFAIR you said that even though Glue API docs say that some fields are "not required" not setting them when making the API call causes an error. Correct?

GaruGaru replied:
I added a comment in the discussion about this issue

hashhar commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/glue/TestHiveGlueMetastore.java:
```diff
@@ -152,34 +158,22 @@ public void testRenameTable()
+            throws Exception
     {
-        // column statistics are not supported by Glue
+        assertThatThrownBy(super::testUpdatePartitionColumnStatisticsEmptyOptionalFields)
```
Let's add some comment about why this is failing? AFAIR you said that even though Glue API docs say that some fields are "not required" not setting them when making the API call causes an error. Correct?

GaruGaru replied:
I added a comment in the discussion about this issue

hashhar commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/AbstractTestHive.java:
```diff
@@ -513,23 +513,23 @@ private static RowType toRowType(List<ColumnMetadata> columns)
                             .put("t_varchar", createStringColumnStatistics(OptionalLong.of(100), OptionalLong.of(230), OptionalLong.of(5), OptionalLong.of(3)))
-                            .put("t_char", createStringColumnStatistics(OptionalLong.of(5), OptionalLong.of(500), OptionalLong.of(1), OptionalLong.of(4)))
-                            .put("t_varbinary", createBinaryColumnStatistics(OptionalLong.of(4), OptionalLong.of(300), OptionalLong.of(1)))
+                            .put("t_char", createStringColumnStatistics(OptionalLong.of(5), OptionalLong.of(50), OptionalLong.of(1), OptionalLong.of(4)))
```
Why this change? This indicates to me that are we somehow calculating data size differently for Thrift vs Glue?

GaruGaru replied:
Here I'm using the `ThriftMetastoreUtil#getAverageColumnLength` function I had to change the values because Glue  reject logically inconsistent statistics.
In particular:

**t_varbinary**
```
test data: 
 rows:  15
 nulls: 1
 max_size:   4
 total_size: 300

avg size: 300/(15rows - 1null) = 20...  > 4 (max)  # ERROR 
```

**t_char**
```
 rows: 15
 nulls: 1 
 max_size: 5
 total_size: 500

avg size: 500/(15rows - 1null) = 35... > 5 (max) # ERROR 
```

losipiuk replied:
> because Glue reject logically inconsistent statistics

Are we sure that statistics computation mechanism used by Tino will not trigger this safety valve?
cc: @findepi

hashhar commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/AbstractTestHive.java:
```diff
@@ -545,10 +545,10 @@ private static RowType toRowType(List<ColumnMetadata> columns)
                             .put("t_bigint", createIntegerColumnStatistics(OptionalLong.of(2345L), OptionalLong.of(6789L), OptionalLong.of(4), OptionalLong.of(7)))
                             .put("t_integer", createIntegerColumnStatistics(OptionalLong.of(234L), OptionalLong.of(678L), OptionalLong.of(5), OptionalLong.of(6)))
                             .put("t_smallint", createIntegerColumnStatistics(OptionalLong.of(23L), OptionalLong.of(65L), OptionalLong.of(7), OptionalLong.of(5)))
-                            .put("t_tinyint", createIntegerColumnStatistics(OptionalLong.of(12), OptionalLong.of(3L), OptionalLong.of(2), OptionalLong.of(3)))
```
@losipiuk does it make sense for a follow-up task in future to normalize such incorrect stats? Or does the engine already ignore/correct for such stats.

cc: @sopel

hashhar replied:
Hmmmm, @GaruGaru I don't think logically incorrect stats value should be a problem. If you changed these values just because they were logically inconsistent, let's revert the change.

GaruGaru replied:
I had to change those values because Glue api reject values that are logically inconsistent ( eg `avg > max`)

hashhar reviewed (commented):
Mostly editorial comments about codestyle, possible NPEs and some broader exception handling to be safe.

Thanks a lot for being patient over reviews and working on this. 🙂

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/DefaultGlueColumnStatisticsProvider.java:
```diff
@@ -0,0 +1,266 @@
+
+            HiveBasicStatistics tableStatistics = getHiveBasicStatistics(table.getParameters());
+            ImmutableMap.Builder<String, HiveColumnStatistics> columnStatsMapBuilder = ImmutableMap.builder();
+            for (CompletableFuture<GetColumnStatisticsForTableResult> future : getStatsFutures) {
```
`getFutureValue(allOf` should work here. But +1 to keep it as a followup.

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/DefaultGlueColumnStatisticsProvider.java:
```diff
@@ -0,0 +1,266 @@
+    {
+        try {
+            List<List<Column>> columnChunks = Lists.partition(partition.getColumns(), GLUE_COLUMN_READ_STAT_PAGE_SIZE);
+            ImmutableList.Builder<CompletableFuture<GetColumnStatisticsForPartitionResult>> getStatsFutures = ImmutableList.builder();
+
+            columnChunks.forEach(partialColumns -> getStatsFutures.add(CompletableFuture.supplyAsync(() -> {
+                List<String> columnsNames = partialColumns.stream()
+                        .map(Column::getName)
+                        .collect(toImmutableList());
+
+                GetColumnStatisticsForPartitionRequest request = new GetColumnStatisticsForPartitionRequest()
+                        .withCatalogId(catalogId)
+                        .withDatabaseName(partition.getDatabaseName())
+                        .withTableName(partition.getTableName())
+                        .withColumnNames(columnsNames)
+                        .withPartitionValues(partition.getValues());
+
+                return glueClient.getColumnStatisticsForPartition(request);
+            }, readExecutor)));
```
nit: 
```suggestion
            List<CompletableFuture<GetColumnStatisticsForPartitionResult>> getStatsFutures = columnChunks.stream()
                    .map(partialColumns -> CompletableFuture.supplyAsync(
                            () -> {
                                List<String> columnsNames = partialColumns.stream()
                                        .map(Column::getName)
                                        .collect(toImmutableList());

                                GetColumnStatisticsForPartitionRequest request = new GetColumnStatisticsForPartitionRequest()
                                        .withCatalogId(catalogId)
                                        .withDatabaseName(partition.getDatabaseName())
                                        .withTableName(partition.getTableName())
                                        .withColumnNames(columnsNames)
                                        .withPartitionValues(partition.getValues());
                                return glueClient.getColumnStatisticsForPartition(request);
                            }, readExecutor))
                    .collect(toImmutableList());
```

@hashhar which one looks nicer to you?

losipiuk replied:
Oh I see that you go this appraach in `getTableColumnStatistics`. Please be consistent.

losipiuk replied:
Still not consistent. `getPartitionColumnStatistics` uses `forEach ... getStatsFutures.add(`. And `getTableColumnStatistics` uses `.collect(toImmutableList());`. Please migrate to `.collect(toImmutableList());` with both.

GaruGaru replied:
Sorry you're right !

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/DefaultGlueColumnStatisticsProvider.java:
```diff
@@ -0,0 +1,266 @@
+
+            getFutureValue(allOf(writeChunkFutures.toArray(CompletableFuture[]::new)));
+
+            final Map<String, HiveColumnStatistics> currentTableColumnStatistics = this.getTableColumnStatistics(table);
```
drop final

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/DefaultGlueColumnStatisticsProvider.java:
```diff
@@ -0,0 +1,266 @@
+            List<ColumnStatistics> columnStats = toGlueColumnStatistics(table, updatedTableColumnStatistics, tableStats.getRowCount());
+            List<List<ColumnStatistics>> columnChunks = Lists.partition(columnStats, GLUE_COLUMN_WRITE_STAT_PAGE_SIZE);
+
+            List<CompletableFuture<Void>> writeChunkFutures = columnChunks.stream().map(columnChunk -> runAsync(
```
is there a reason to do updates first and removals later? It feels to me it would be better to gather the futures on single list and call `getFutureValue(allOf` just once. 
So even in case of an error for some of the calls we have as many changes applied as possible.

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/DefaultGlueColumnStatisticsProvider.java:
```diff
@@ -0,0 +1,266 @@
+
+            List<List<ColumnStatistics>> columnChunks = Lists.partition(columnStats, GLUE_COLUMN_WRITE_STAT_PAGE_SIZE);
+
+            List<CompletableFuture<Void>> writePartitionStatsFutures = columnChunks.stream()
```
same here. Wouldn't it be better to collect futures on single list?

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/GlueHiveMetastore.java:
```diff
@@ -385,8 +397,10 @@ public void updatePartitionStatistics(HiveIdentity identity, Table table, String
+            final Map<String, String> updateStatisticsParameters = updateStatisticsParameters(partition.getParameters(), updatedStatistics.getBasicStatistics());
+            partitionInput.setParameters(updateStatisticsParameters);
+            partition = Partition.builder(partition).setParameters(updateStatisticsParameters).build();
+            columnStatisticsProvider.updatePartitionStatistics(partition, updatedStatistics.getColumnStatistics());
```
please move call to `columnStatisticsProvider.updatePartitionStatistics` below `glueClient.updatePartition` we update the basic stats first.

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/GlueHiveMetastore.java:
```diff
@@ -358,8 +368,10 @@ public void updateTableStatistics(HiveIdentity identity, String databaseName, St
+            final Map<String, String> statisticsParameters = updateStatisticsParameters(table.getParameters(), updatedStatistics.getBasicStatistics());
+            tableInput.setParameters(statisticsParameters);
+            table = Table.builder(table).setParameters(statisticsParameters).build();
+            columnStatisticsProvider.updateTableColumnStatistics(table, updatedStatistics.getColumnStatistics());
```
plase move the call to `columnStatisticsProvider.updateTableColumnStatistics(table, updatedStatistics.getColumnStatistics());`  below `glueClient.updateTable` so we update basic stats first.

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/converter/GlueStatConverter.java:
```diff
@@ -0,0 +1,291 @@
+                        OptionalLong.of(catalogBooleanData.getNumberOfFalses()),
+                        fromMetastoreNullsCount(catalogBooleanData.getNumberOfNulls()));
+
+            case DATE:
+                DateColumnStatisticsData catalogDateData = catalogColumnStatisticsData.getDateColumnStatisticsData();
+                Optional<LocalDate> minOfDate = dateToLocalDate(catalogDateData.getMinimumValue());
+                Optional<LocalDate> maxOfDate = dateToLocalDate(catalogDateData.getMaximumValue());
+                OptionalLong nullsCountOfDate = fromMetastoreNullsCount(catalogDateData.getNumberOfNulls());
+                OptionalLong distinctValuesCountOfDate = OptionalLong.of(catalogDateData.getNumberOfDistinctValues());
+                return createDateColumnStatistics(minOfDate, maxOfDate, nullsCountOfDate, fromMetastoreDistinctValuesCount(distinctValuesCountOfDate, nullsCountOfDate, rowCount));
+
+            case DECIMAL:
+                DecimalColumnStatisticsData catalogDecimalData = catalogColumnStatisticsData.getDecimalColumnStatisticsData();
+                Optional<BigDecimal> minOfDecimal = glueDecimalToBigDecimal(catalogDecimalData.getMinimumValue());
+                Optional<BigDecimal> maxOfDecimal = glueDecimalToBigDecimal(catalogDecimalData.getMaximumValue());
+                OptionalLong distinctValuesCountOfDecimal = OptionalLong.of(catalogDecimalData.getNumberOfDistinctValues());
+                OptionalLong nullsCountOfDecimal = fromMetastoreNullsCount(catalogDecimalData.getNumberOfNulls());
+                return createDecimalColumnStatistics(minOfDecimal, maxOfDecimal, nullsCountOfDecimal, fromMetastoreDistinctValuesCount(distinctValuesCountOfDecimal, nullsCountOfDecimal, rowCount));
```
Use braces. This will allow you to use shorter variable names. Like this:
```suggestion
            case DATE: {
                DateColumnStatisticsData data = catalogColumnStatisticsData.getDateColumnStatisticsData();
                Optional<LocalDate> min = dateToLocalDate(data.getMinimumValue());
                Optional<LocalDate> max = dateToLocalDate(data.getMaximumValue());
                OptionalLong nullsCount = fromMetastoreNullsCount(data.getNumberOfNulls());
                OptionalLong numberOfDistinctValues = OptionalLong.of(data.getNumberOfDistinctValues());
                return createDateColumnStatistics(min, max, nullsCount, fromMetastoreDistinctValuesCount(numberOfDistinctValues, nullsCount, rowCount));
            }
            case DECIMAL: {
                DecimalColumnStatisticsData data = catalogColumnStatisticsData.getDecimalColumnStatisticsData();
                Optional<BigDecimal> min = glueDecimalToBigDecimal(data.getMinimumValue());
                Optional<BigDecimal> max = glueDecimalToBigDecimal(data.getMaximumValue());
                OptionalLong numberOfDistinctValues = OptionalLong.of(data.getNumberOfDistinctValues());
                OptionalLong nullsCount = fromMetastoreNullsCount(data.getNumberOfNulls());
                return createDecimalColumnStatistics(min, max, nullsCount, fromMetastoreDistinctValuesCount(numberOfDistinctValues, nullsCount, rowCount));
            }
```

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/converter/GlueStatConverter.java:
```diff
@@ -0,0 +1,291 @@
+        ColumnStatisticsData catalogColumnStatisticsData = new ColumnStatisticsData();
+
+        switch (((PrimitiveTypeInfo) typeInfo).getPrimitiveCategory()) {
+            case BOOLEAN:
```
use braces here to to have separate namespace for each type considered.

losipiuk reviewed (commented):
Thanks @GaruGaru. Awesome work! Please address remaining comments and I think we are good to merge.

Also thanks @hashhar for very thorough review!

GaruGaru commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/converter/GlueStatConverter.java:
```diff
@@ -0,0 +1,291 @@
+
+    private static Date localDateToDate(LocalDate date)
+    {
+        long millisecondsSinceEpoch = date.toEpochDay() * MILLIS_PER_DAY;
```
the `ThriftMetastoreUtil#toMetastoreDate` doesn't look like it can handle null references for its argument since it calls `arg.toEpochDay()`. 
How do you think we should handle this particular case ?

hashhar replied:
That method doesn't need to check since the caller only calls it if the Optional has a value.

I think you are safe too since this method isn't ever called with a possible null value. Consider this resolved.

GaruGaru commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/converter/GlueStatConverter.java:
```diff
@@ -0,0 +1,291 @@
+
+    private static Optional<LocalDate> dateToLocalDate(Date date)
+    {
+        long daysSinceEpoch = date.getTime() / MILLIS_PER_DAY;
```
Do you think that we can use `ThriftMetastoreUtil` here or is better to  keep the implementations as separated as possibile ?

hashhar replied:
I think it's safe to do so. We already use ThriftMetastoreUtil in GlueHiveMetastore for other things.

GaruGaru commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/converter/GlueStatConverter.java:
```diff
@@ -0,0 +1,291 @@
+                return createBinaryColumnStatistics(
+                        maxColumnLengthOfBinary,
+                        getTotalSizeInBytes(averageColumnLengthOfBinary, rowCount, nullsCountOfBinary),
+                        nullsCountOfBinary);
```
the `nullsCountOfBinary` variable is the result of `fromMetastoreNullsCount(catalogBinaryData.getNumberOfNulls());` in this method is assigned to a variable because the value is used twice

GaruGaru commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/converter/GlueStatConverter.java:
```diff
@@ -0,0 +1,291 @@
+
+    private static DecimalNumber bigDecimalToGlueDecimal(BigDecimal decimal)
+    {
+        Decimal hiveDecimal = new Decimal((short) decimal.scale(), ByteBuffer.wrap(decimal.unscaledValue().toByteArray()));
```
Agree but in this cases the method is called inside an `Optional#isPresent` lambda, NPEs should not happen.
In order to be as defensive as possible do you think we can return an `Optional` here and set min/max when present ?

hashhar replied:
I missed the call-site when reviewing. This should be fine. Please mark other such instances as resolved. Sorry for the false alarms.

GaruGaru commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/converter/GlueStatConverter.java:
```diff
@@ -0,0 +1,291 @@
+    private static ColumnStatisticsData toGlueColumnStatisticsData(HiveColumnStatistics statistics, HiveType columnType, OptionalLong rowCount)
+    {
+        TypeInfo typeInfo = columnType.getTypeInfo();
+        checkArgument(typeInfo.getCategory() == PRIMITIVE, "unsupported type: %s", columnType);
```
Should we also add some context to the error ? eg `Unsupported statistics type: %s, Only primitive types are supported`

hashhar replied:
I don't think users would know what a PrimitiveType is as defined by the metastore. I'll defer on this to @trinodb/maintainers .

GaruGaru commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/converter/GlueStatConverter.java:
```diff
@@ -0,0 +1,291 @@
+            case SHORT:
+            case INT:
+            case LONG:
+            case TIMESTAMP:
```
The intent here is to handle all the types (`SHORT,INT,LONG,TIMESTAMP`) Using the same logic by leveraging java `switch` multiple cases, so in this case all the defined types are written with the `LongColumnStatisticsData` object to glue, Glue doesn't support short, int and timestamp values. 

Should we make this more explicit ? 
Another possible change is to handle `timestamp` type as `Date` instead of `long` but I'm not sure if this can help

GaruGaru commented:
Thank you for the time and effort that you are putting in reviewing the code deeply!
I  have addressed most of the comments. 

I'll investigate the Glue's behaviour in  `testUpdateTableColumnStatisticsEmptyOptionalFields` and `testUpdatePartitionColumnStatisticsEmptyOptionalFields`.
 
Also: Should we implement some kind of rollback strategy for Glue stats partial write/delete failures ? 
Can having partially updated statistics impact user experience significally ?

hashhar commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/DefaultGlueColumnStatisticsProvider.java:
```diff
@@ -156,7 +150,10 @@ public DefaultGlueColumnStatisticsProvider(AWSGlueAsync glueClient, String catal
             return columnStatsMapBuilder.build();
         }
-        catch (AmazonServiceException ex) {
+        catch (EntityNotFoundException ex) {
```
@losipiuk should we fail if Trino tries to fetch statistics for a partition which doesn't exist? Or return empty stats?

hashhar replied:
Btw, this was changed to avoid catching the `TrinoException(HIVE_METASTORE_ERROR, EntityNotFoundException)` in the caller to detect that a partition doesn't exist.

I'd prefer if this responsibility of handling `EntityNotFoundException` would be given to the caller since this is a public method and part of the API.

WDYT @losipiuk ?

hashhar commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/DefaultGlueColumnStatisticsProvider.java:
```diff
@@ -169,34 +166,32 @@ public void updateTableColumnStatistics(Table table, Map<String, HiveColumnStati
                                             .withTableName(table.getTableName())
-                                            .withColumnName(column)))
-                    ).collect(toUnmodifiableList());
+                                            .withColumnName(column)), this.writeExecutor)).collect(toUnmodifiableList());
```
nit: move `collect` to new line.

hashhar commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/DefaultGlueColumnStatisticsProvider.java:
```diff
@@ -169,34 +166,32 @@ public void updateTableColumnStatistics(Table table, Map<String, HiveColumnStati
-            for (CompletableFuture<Void> writeFuture : writeFutures) {
-                getFutureValue(writeFuture);
-            }
+            final ImmutableList.Builder<CompletableFuture<Void>> updateOperationsFutures = ImmutableList.builderWithExpectedSize(updateFutures.size() + deleteFutures.size());
```
drop final.

hashhar commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/DefaultGlueColumnStatisticsProvider.java:
```diff
@@ -217,45 +212,29 @@ public void updatePartitionStatistics(Partition partition, Map<String, HiveColum
                                     .withPartitionValues(partition.getValues())
-                                    .withColumnStatisticsList(columnChunk)), writeExecutor)
-                    ).collect(toUnmodifiableList());
+                                    .withColumnStatisticsList(columnChunk)), writeExecutor)).collect(toUnmodifiableList());
```
nit: move `collect` to new line.

hashhar commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/DefaultGlueColumnStatisticsProvider.java:
```diff
@@ -217,45 +212,29 @@ public void updatePartitionStatistics(Partition partition, Map<String, HiveColum
                                     .withPartitionValues(partition.getValues())
-                                    .withColumnName(column)), writeExecutor)
-                    ).collect(toUnmodifiableList());
+                                    .withColumnName(column)), writeExecutor)).collect(toUnmodifiableList());
```
nit: move `collect` to new line.

hashhar commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/DefaultGlueColumnStatisticsProvider.java:
```diff
@@ -217,45 +212,29 @@ public void updatePartitionStatistics(Partition partition, Map<String, HiveColum
-            for (CompletableFuture<Void> writeFuture : writeFutures) {
-                getFutureValue(writeFuture);
-            }
+            final ImmutableList.Builder<CompletableFuture<Void>> updateOperationsFutures = ImmutableList.builderWithExpectedSize(writePartitionStatsFutures.size() + deleteStatsFutures.size());
```
drop final.

hashhar reviewed (approved):
Looks good to go % some editorial nitpicking and the `EntityNotFoundException` handling in `getPartitionColumnStatistics` (https://github.com/trinodb/trino/pull/6178#discussion_r571689018).

Thanks again for the work you've put into this @GaruGaru . ❤️

losipiuk commented on plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/glue/DefaultGlueColumnStatisticsProvider.java:
```diff
@@ -0,0 +1,245 @@
+
+    private List<String> getAllColumns(Table table)
+    {
+        final ImmutableList.Builder<String> allColumns = ImmutableList.builderWithExpectedSize(table.getDataColumns().size() + table.getPartitionColumns().size());
```
drop final

losipiuk reviewed (commented):
Thanks. Looks mostly good.
Some nit picks and two important questions:

* https://github.com/trinodb/trino/pull/6178#discussion_r571961702
* https://github.com/trinodb/trino/pull/6178#discussion_r571955866

GaruGaru commented:
@findepi @hashhar I did some further investigation for the `testUpdateTableColumnStatisticsEmptyOptionalFields` and `testUpdatePartitionColumnStatisticsEmptyOptionalFields` tests: 

The tests are failing when passing empty column statistics with `Internal Server Error 500` Glue.

I narrowed down the issue and it looks like the problem is only present with `DecimalColumnStatisticsData` statistics having min/max  `null`.

Looking at the [docs for DecimalColumnStatisticsData](https://docs.aws.amazon.com/glue/latest/webapi/API_DecimalColumnStatisticsData.html)  `min`, `max` fields are not required. 

But looking at the [docs for inner object DecimalNumber](https://docs.aws.amazon.com/glue/latest/webapi/API_DecimalNumber.html) `Scale` and `UnscaledValue`  are required, could it be a bug in the glue api ? 

In particular the statistics that are making Glue fail are: 
```java
.put("t_short_decimal", createDecimalColumnStatistics(Optional.empty(), Optional.empty(), OptionalLong.of(5), OptionalLong.of(7)))
.put("t_long_decimal", createDecimalColumnStatistics(Optional.empty(), Optional.empty(), OptionalLong.of(2), OptionalLong.of(1)))
```

losipiuk commented:
> @findepi @hashhar I did some further investigation for the `testUpdateTableColumnStatisticsEmptyOptionalFields` and `testUpdatePartitionColumnStatisticsEmptyOptionalFields` tests:
> 
> The tests are failing when passing empty column statistics with `Internal Server Error 500` Glue.
> 
> I narrowed down the issue and it looks like the problem is only present with `DecimalColumnStatisticsData` statistics having min/max `null`.
> 
> Looking at the [docs for DecimalColumnStatisticsData](https://docs.aws.amazon.com/glue/latest/webapi/API_DecimalColumnStatisticsData.html) `min`, `max` fields are not required.
> 
> But looking at the [docs for inner object DecimalNumber](https://docs.aws.amazon.com/glue/latest/webapi/API_DecimalNumber.html) `Scale` and `UnscaledValue` are required, could it be a bug in the glue api ?
> 
> In particular the statistics that are making Glue fail are:
> 
> ```java
> .put("t_short_decimal", createDecimalColumnStatistics(Optional.empty(), Optional.empty(), OptionalLong.of(5), OptionalLong.of(7)))
> .put("t_long_decimal", createDecimalColumnStatistics(Optional.empty(), Optional.empty(), OptionalLong.of(2), OptionalLong.of(1)))
> ```
> 
> Am I missing something here ? Should we think about a workaround for this ?

Thanks a lot for the investigation. 
This indeed seems like a problem with Glue API (or documentation). This could also be problem with client SDK but I doubt bumping it would help as we are at almost newest version.

Can you, as a workaround, generate min/max values for statistics for decimal columns putting smallest and biggest value for a type if the statistic is Optional.empty()?

GaruGaru commented:
@losipiuk Thank you for the suggestion that should work, but the value is actually of `BigDecimal` type, I'm not sure about using MIN / MAX.
I've made some tests using `Double.MAX_VALUE` and `Double.MIN_VALUE` and things are working as expected.

I've also found another interesting aws glue sdk behaviour that I need to investigate before applying this chages:

Null values for  `min/max` stats writes for a partition/column are performed correctly, but the subsequent reads for the same min/max stats returns `0` instead of an empty value such as `null`.
I don't have much understanding about how Trino uses these statistics but `0` looks like a sensible value when checking min/max bounds.

losipiuk commented:
> @losipiuk Thank you for the suggestion that should work, but the value is actually of `BigDecimal` type, I'm not sure about using MIN / MAX.
> I've made some tests using `Double.MAX_VALUE` and `Double.MIN_VALUE` and things are working as expected.

Cool. I was thinking of deriving min/max value from the actual SQL type. So for DECIMAL(10,2) the min value would be `-9999999,99` and max value `99999999,99`.

> 
> I've also found another interesting aws glue sdk behaviour that I need to investigate before applying this changes:
> 
> When writing null values for `min/max` statistics the subsequent reads from Glue will return `0` instead of `null`.
> I don't have much understanding about how Trino uses these statistics but `0` looks like a sensible value when checking min/max bounds.

What do you mean by writing `null` as min/max? The case when we do not have value for those and we are not putting the value for min/max in the PUT request?
This is suboptimal. If that is really the behaviour of Glue we should craft fake min/max values not only for DECIMAL columns, but for all columns when we do not have proper values.

findepi commented:
> Cool. I was thinking of deriving min/max value from the actual SQL type. So for DECIMAL(10,2) the min value would be `-9999999,99` and max value `99999999,99`.

This would work, but is not very simple. What are the cases when min/max is missing but other stats are not?
(ie Can we afford ignoring all stats in such case?)

losipiuk commented:
> > Cool. I was thinking of deriving min/max value from the actual SQL type. So for DECIMAL(10,2) the min value would be `-9999999,99` and max value `99999999,99`.
> 
> This would work, but is not very simple. What are the cases when min/max is missing but other stats are not?
> (ie Can we afford ignoring all stats in such case?)

That is a very good question. I skimmed the code and it looks that if the data type is supported for min/max we will not set the values only for the case that all values in column are NULL (or NULL/NaN for DOUBLE/REAL columns). Probably not super typical case. WDYT?

findepi commented:
If this is about NULLs and NaNs only, let's keep things simple.

GaruGaru commented:
I've pushed a possible (ugly) workaround using min/max values to discriminate between statistics that are not present and actually min/max with value `0`.

When writing if min/max is null the min/max value for the type will be written.
When reading if min/max value is equals to the min/max value for the type then an empty value will be returned. 

All tests are passing now.

losipiuk commented:
> I've pushed a possible (ugly) workaround using min/max values to discriminate between statistics that are not present and actually min/max with value `0`.
> 
> When writing if min/max is null the min/max value for the type will be written.
> When reading if min/max value is equals to the min/max value for the type then an empty value will be returned.

It makes semantics quirky. You can read `Optional.empty()` for valid statistics, which by chance happen to actually contain the marker value.
I think it would be better (and simpler) to follow the suggestion from @findepi (https://github.com/trinodb/trino/pull/6178#issuecomment-781331579) to just not write statistics for a column at all if min/max are not present.

GaruGaru commented:
I've updated the implementation to skip write for statistics that have min or max null

losipiuk commented:
> I've updated the implementation to skip write for statistics that have min or max null

Thanks. Sorry for delay, this is great. 
One final request. Can be a separate PR. Can we have a test which would tell us when AWS fixed their API and we can actually send requests with missing min/max.

I restarted the CI. I think this is mergable when green.

losipiuk commented:
squashed
